qiskit_dynamics/solvers/solver_classes.py

```diff
@@ -547,87 +551,63 @@ def format_final_states(y, model, y_input, y0_cls):
+    t_span_input: Array,
+    y0_input: Union[Array, QuantumState, BaseOperator],
+    signals_input: Optional[Union[List[Signal], Tuple[List[Signal], List[Signal]]]],
 ) -> Tuple[List, List, List, bool]:
-    """Helper function for setting up lists of simulations.
-
-    Transform input signals, t_span, and y0 into lists of the same length.
-    Arguments are given as either lists of valid specifications, or as a singleton of a valid
-    specification. Singletons are transformed into a list of length one, then all arguments
-    are expanded to be the same length as the longest argument max_len:
-        - If len(arg) == 1, it will be repeated max_len times
-        - if len(arg) == max_len, nothing is done
-        - If len(arg) not in (1, max_len), an error is raised
-
-    Args:
-        t_span: Time interval specification.
-        y0: Initial state specification.
-        signals: Signal specification.
-
-    Returns:
-        Tuple: tuple of lists of arguments of the same length, along with a bool specifying whether
-        the arguments specified multiple simulations or not.
-
-    Raises:
-        QiskitError: If the length of any arguments are incompatible, or if any singleton
-        is an invalid shape.
+    """Setup t_span, y0, and signals for doing a list of simulations. Utilizes
+    solver_utils.setup_arg_lists.
     """
 
-    multiple_sims = False
-
-    if signals is None:
-        signals = [signals]
-    elif isinstance(signals, tuple):
-        # single Lindblad
-        signals = [signals]
-    elif isinstance(signals, list) and isinstance(signals[0], tuple):
-        # multiple lindblad
-        multiple_sims = True
-    elif isinstance(signals, list) and isinstance(signals[0], (list, SignalList)):
-        # multiple Hamiltonian signals lists
-        multiple_sims = True
-    elif isinstance(signals, SignalList) or (
-        isinstance(signals, list) and not isinstance(signals[0], (list, SignalList))
-    ):
-        # single Hamiltonian signals list
-        signals = [signals]
-    else:
-        raise QiskitError("Signals specified in invalid format.")
+    def signals_to_list(signals):
+        was_list = False
+        if signals is None:
+            signals = [signals]
+        elif isinstance(signals, tuple):
+            # single Lindblad
+            signals = [signals]
+        elif isinstance(signals, list) and isinstance(signals[0], tuple):
+            # multiple lindblad
+            was_list = True
+        elif isinstance(signals, list) and isinstance(signals[0], (list, SignalList)):
+            # multiple Hamiltonian signals lists
+            was_list = True
+        elif isinstance(signals, SignalList) or (
+            isinstance(signals, list) and not isinstance(signals[0], (list, SignalList))
+        ):
+            # single Hamiltonian signals list
+            signals = [signals]
+        else:
+            raise QiskitError("Signals specified in invalid format.")
```

Comment:
You could combine some of these, but I like it better as is, I think it's clearer visually